src/device/usbd.c

```diff
@@ -758,9 +771,13 @@ static bool process_get_descriptor(uint8_t rhport, tusb_control_request_t const
       uint16_t len = sizeof(tusb_desc_device_t);
 
       // Only send up to EP0 Packet Size if not addressed
+      // This only happens with the very first get device descriptor and EP0 size = 8 or 16.
       if ((CFG_TUD_ENDPOINT0_SIZE < sizeof(tusb_desc_device_t)) && !_usbd_dev.addressed)
       {
         len = CFG_TUD_ENDPOINT0_SIZE;
+
+        // Hack here: we modify the request length to prevent usbd_control response with zlp
+        ((tusb_control_request_t*) p_request)->wLength = CFG_TUD_ENDPOINT0_SIZE;

```

Comment:
@pigrew @cr1901 this is a hack to modify wLength, else usbd_control will send out 8 byte + ZLP. This is one time hack and only needed for msp430 (and probably 8-bit mcu), I don't know anyway cleaner.

Reply:
@hathach My [solution](https://github.com/cr1901/tinyusb/commit/aef7661ea8a321de28ed6dd09e2bad82c41a927b#diff-e546d6e4c4e6a5c18b285fbe07514f89) to #184 does basically the same thing as you, except setting `wLength` directly like you do (since that takes priority if smaller than `len`) is probably the better solution.

If it isn't clean, I think we'll live for now :P.